conda/gateways/repodata/jlap/fetch.py

```diff
@@ -0,0 +1,402 @@
+# Copyright (C) 2012 Anaconda, Inc
+# SPDX-License-Identifier: BSD-3-Clause
+# Lappin' up the jlap
+from __future__ import annotations
+import io
+
+import json
+import logging
+import pathlib
+import pprint
+import re
+import time
+import zstandard
+from contextlib import contextmanager
+from hashlib import blake2b
+from typing import Iterator
+
+import jsonpatch
+from requests import HTTPError
+
+from conda.base.context import context
+from conda.gateways.connection import Response, Session
+from conda.gateways.repodata import RepodataState
+
+from .core import JLAP
+
+log = logging.getLogger(__name__)
+
+
+DIGEST_SIZE = 32  # 160 bits a minimum 'for security' length?
+
+JLAP_KEY = "jlap"
+HEADERS = "headers"
+NOMINAL_HASH = "nominal_hash"
+ON_DISK_HASH = "actual_hash"
+LATEST = "latest"
+JLAP_UNAVAILABLE = "jlap_unavailable"
+ZSTD_UNAVAILABLE = "zstd_unavailable"
+
+
+def hash():
+    """
+    Ordinary hash.
+    """
+    return blake2b(digest_size=DIGEST_SIZE)
+
+
+def get_place(url, extra=""):
+    if "current_repodata" in url:
+        extra = f".c{extra}"
+    return pathlib.Path("-".join(url.split("/")[-3:-1])).with_suffix(f"{extra}.json")
+
+
+class Jlap304NotModified(Exception):
+    pass
+
+
+class JlapSkipZst(Exception):
+    pass
+
+
+class JlapPatchNotFound(LookupError):
+    pass
+
+
+def process_jlap_response(response: Response, pos=0, iv=b""):
+    # if response is 304 Not Modified, could return a buffer with only the
+    # cached footer...
+    if response.status_code == 304:
+        raise Jlap304NotModified()
+
+    def lines() -> Iterator[bytes]:
+        yield from response.iter_lines(delimiter=b"\n")  # type: ignore
+
+    buffer = JLAP.from_lines(lines(), iv, pos)
+
+    # new iv == initial iv if nothing changed
+    pos, footer, _ = buffer[-2]
```

Comment:
All of these buffer indices (`-1`, `-2`, and `-3`), are effectively magic numbers as it isn't immediately clear what these indices are referring to. Since the buffer object is an instance of a custom list implementation can we create properties instead that return this information in a way that is more descriptive?

Reply:
Like https://github.com/conda/conda/pull/12090/commits/4e70c5e4168342d8610be0dbd091005a25f91d16 ?

Reply:
It is standard Python that -1 is the last element of the list, -2 is the next to last and so on...?